File: backend/src/main/java/com/pm/backend/controller/NotificationController.java
package com.pm.backend.controller;

import com.pm.backend.model.Notification;
import com.pm.backend.model.User;
import com.pm.backend.repository.NotificationRepository;
import com.pm.backend.repository.UserRepository;
import com.pm.backend.security.UserPrincipal;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import com.pm.backend.service.EmailService;

import java.util.List;

@RestController
@RequestMapping("/api/notifications")
public class NotificationController {

    private final NotificationRepository notificationRepository;
    private final UserRepository userRepository;
    private final EmailService emailService;

    public NotificationController(NotificationRepository notificationRepository, UserRepository userRepository, EmailService emailService) {
        this.notificationRepository = notificationRepository;
        this.userRepository = userRepository;
        this.emailService = emailService;
    }

    @GetMapping
    public ResponseEntity<List<Notification>> getNotifications(@AuthenticationPrincipal UserPrincipal userPrincipal) {
        User recipient = userRepository.findById(userPrincipal.getId()).orElseThrow();
        List<Notification> notifications = notificationRepository.findByRecipientOrderByCreatedAtDesc(recipient);
        return ResponseEntity.ok(notifications);
    }

    @GetMapping("/unread")
    public ResponseEntity<List<Notification>> getUnreadNotifications(@AuthenticationPrincipal UserPrincipal userPrincipal) {
        User recipient = userRepository.findById(userPrincipal.getId()).orElseThrow();
        List<Notification> notifications = notificationRepository.findByRecipientAndIsReadFalseOrderByCreatedAtDesc(recipient);
        return ResponseEntity.ok(notifications);
    }

    @PutMapping("/{id}/read")
    public ResponseEntity<Notification> markAsRead(@PathVariable Long id, @AuthenticationPrincipal UserPrincipal userPrincipal) {
        Notification notification = notificationRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Notification not found: " + id));

        if (!notification.getRecipient().getId().equals(userPrincipal.getId())) {
            return ResponseEntity.status(403).build();
        }

        notification.setIsRead(true);
        Notification updatedNotification = notificationRepository.save(notification);
        return ResponseEntity.ok(updatedNotification);
    }

    @PutMapping("/read-all")
    public ResponseEntity<?> markAllAsRead(@AuthenticationPrincipal UserPrincipal userPrincipal) {
        User recipient = userRepository.findById(userPrincipal.getId()).orElseThrow();
        List<Notification> unread = notificationRepository.findByRecipientAndIsReadFalseOrderByCreatedAtDesc(recipient);
        for (Notification n : unread) {
            n.setIsRead(true);
        }
        notificationRepository.saveAll(unread);
        return ResponseEntity.ok().build();
    }

    @PostMapping
    public ResponseEntity<?> createNotification(@RequestBody java.util.Map<String, Object> payload,
                                               @AuthenticationPrincipal UserPrincipal userPrincipal) {
        String title = (String) payload.getOrDefault("title", "Notification Alert");
        String message = (String) payload.getOrDefault("message", "System update");
        String type = (String) payload.getOrDefault("type", "SYSTEM_ALERT");
        Object recipientIdObj = payload.get("recipientId");

        List<User> recipients;
        if (recipientIdObj != null && !"ALL".equals(String.valueOf(recipientIdObj))) {
            try {
                Long recipientId = Long.parseLong(String.valueOf(recipientIdObj));
                recipients = userRepository.findById(recipientId).map(List::of).orElseGet(() -> userRepository.findAll());
            } catch (Exception e) {
                recipients = userRepository.findAll();
            }
        } else {
            recipients = userRepository.findAll();
        }

        for (User recipient : recipients) {
            Notification notification = Notification.builder()
                    .title(title)
                    .message(message)
                    .type(type)
                    .isRead(false)
                    .recipient(recipient)
                    .build();
            notificationRepository.save(notification);

            if (recipient.getEmail() != null && recipient.getEmail().contains("@")) {
                emailService.sendEmailAlert(recipient.getEmail(), "[Project Workspace Alert] " + title, message);
            }
        }

        return ResponseEntity.ok().body(java.util.Map.of("message", "Notifications dispatched to " + recipients.size() + " user(s)."));
    }
}

File: backend-node/src/controllers/notificationController.ts
import { Response } from 'express';
import { Notification, User } from '../models';
import { AuthRequest } from '../middleware/auth';
import { EmailService } from '../services/emailService';

export class NotificationController {
  public static async getNotifications(req: AuthRequest, res: Response) {
    try {
      if (!req.user) return res.status(401).send('Unauthorized');

      const notifications = await Notification.findAll({
        where: { recipientId: req.user.id },
        order: [['createdAt', 'DESC']],
      });
      return res.json(notifications);
    } catch (err: any) {
      console.error('Error in getNotifications:', err);
      return res.status(500).json({ error: err.message });
    }
  }

  public static async getUnreadNotifications(req: AuthRequest, res: Response) {
    try {
      if (!req.user) return res.status(401).send('Unauthorized');

      const unread = await Notification.findAll({
        where: {
          recipientId: req.user.id,
          isRead: false,
        },
        order: [['createdAt', 'DESC']],
      });
      return res.json(unread);
    } catch (err: any) {
      console.error('Error in getUnreadNotifications:', err);
      return res.status(500).json({ error: err.message });
    }
  }

  public static async markAsRead(req: AuthRequest, res: Response) {
    try {
      if (!req.user) return res.status(401).send('Unauthorized');

      const id = parseInt(req.params.id);
      const notification = await Notification.findByPk(id);

      if (!notification) return res.status(404).send('Notification not found');

      if (notification.recipientId !== req.user.id) {
        return res.status(403).send('Forbidden');
      }

      notification.isRead = true;
      await notification.save();

      return res.json(notification);
    } catch (err: any) {
      console.error('Error in markAsRead:', err);
      return res.status(500).json({ error: err.message });
    }
  }

  public static async markAllAsRead(req: AuthRequest, res: Response) {
    try {
      if (!req.user) return res.status(401).send('Unauthorized');

      await Notification.update(
        { isRead: true },
        { where: { recipientId: req.user.id, isRead: false } }
      );

      return res.json({ success: true });
    } catch (err: any) {
      console.error('Error in markAllAsRead:', err);
      return res.status(500).json({ error: err.message });
    }
  }

  public static async createNotification(req: AuthRequest, res: Response) {
    try {
      const { title, message, type, recipientId } = req.body;

      let recipients: User[] = [];
      if (recipientId && String(recipientId) !== 'ALL') {
        const singleUser = await User.findByPk(Number(recipientId));
        recipients = singleUser ? [singleUser] : await User.findAll();
      } else {
        recipients = await User.findAll();
      }

      for (const recipient of recipients) {
        await Notification.create({
          title: title || 'Notification Alert',
          message: message || 'System update',
          type: type || 'SYSTEM_ALERT',
          isRead: false,
          recipientId: recipient.id,
        });

        if (recipient.email && recipient.email.includes('@')) {
          EmailService.sendEmailAlert(recipient.email, `[Project Workspace Alert] ${title}`, message || '');
        }
      }

      return res.json({ message: `Notifications dispatched to ${recipients.length} user(s).` });
    } catch (err: any) {
      console.error('Error in createNotification:', err);
      return res.status(500).json({ error: err.message });
    }
  }
}
